File: api/app/models/item.py
# Path: api/app/models/item.py
from sqlalchemy import Column, Integer, String, Text, Float, Enum as SQLAlchemyEnum, JSON # <--- ADD JSON HERE
from sqlalchemy.orm import relationship
from app.db.base_class import Base
from typing import TYPE_CHECKING
import enum # For Python enums

if TYPE_CHECKING:
    from .character_item import CharacterItem # For relationship type hinting

# Python Enum for ItemType
class ItemTypeEnum(enum.Enum):
    WEAPON = "weapon"
    ARMOR = "armor"
    SHIELD = "shield"
    ADVENTURING_GEAR = "adventuring_gear"
    TOOL = "tool"
    POTION = "potion"
    SCROLL = "scroll"
    WAND = "wand"
    RING = "ring"
    WONDROUS_ITEM = "wondrous_item"
    CURRENCY = "currency"
    OTHER = "other"

class Item(Base):
    __tablename__ = "items"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    name = Column(String(100), unique=True, nullable=False, index=True)
    description = Column(Text, nullable=True)
    
    item_type = Column(SQLAlchemyEnum(ItemTypeEnum, name="itemtypeenum"), nullable=False, default=ItemTypeEnum.OTHER)
    
    weight = Column(Float, nullable=True, default=0.0) 
    cost_gp = Column(Float, nullable=True, default=0.0) 

    properties = Column(JSON, nullable=True) # JSON type is now imported

    character_associations = relationship("CharacterItem", back_populates="item_definition")

File: aethoria-ui/src/types/item.ts
// Path: src/types/item.ts

// Using a string literal union type for type safety
export type FrontendItemType = 
    | "weapon"
    | "armor"
    | "shield"
    | "adventuring_gear"
    | "tool"
    | "potion"
    | "scroll"
    | "wand"
    | "ring"
    | "wondrous_item"
    | "currency"
    | "other";

// Optional: If you need an object to iterate over values (e.g., for a dropdown)
export const FrontendItemTypeEnumObject = {
    WEAPON: "weapon",
    ARMOR: "armor",
    SHIELD: "shield",
    ADVENTURING_GEAR: "adventuring_gear",
    TOOL: "tool",
    POTION: "potion",
    SCROLL: "scroll",
    WAND: "wand",
    RING: "ring",
    WONDROUS_ITEM: "wondrous_item",
    CURRENCY: "currency",
    OTHER: "other"
} as const; // 'as const' makes keys readonly and values literal types

export interface ItemDefinition {
  id: number;
  name: string;
  description?: string | null;
  item_type: FrontendItemType; // Use the string literal union type
  weight?: number | null;
  cost_gp?: number | null;
  properties?: any | null; 
}

export interface CharacterItem {
  id: number; 
  item_id: number;
  quantity: number;
  is_equipped: boolean;
  item_definition: ItemDefinition; 
}